keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1525 @@
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+The proposal is that a plugin handles all operations that are specific
+to the resources managed by that plugin. This includes operations at
+the control plane level (tracking where in the cluster resources are
+available, helping with pod scheduling decisions, allocating resources
+when requested) as well as the node level (preparing container
+startup). Such a plugin can be implemented in arbitrary programming
+languages as long as it supports the resource allocation protocol and
+gRPC interfaces defined in this KEP. An utility package with Go
+support code will be made available to simplify the development of
+such a plugin, but using it will not be required and its API is not
+part of this KEP.
+
+Three new API object types get added in a new API group:
+- ResourcePlugin, not namespaced, with a description of the plugin.
+- ResourceClass, not namespaced, with privileged parameters for
+  multiple resource instances of a certain kind. All these instances
+  are provided by the same resource plugin, which is identified by a
+  field in the class.
+- ResourceClaim, namespaced, with parameters provided by a normal user
+  that describes a resource instance that needs to be allocated. A
+  ResourceClaim contains the usual meta data, a spec and a status. The
+  spec identifies the plugin that handles the resource via a class
+  name.
+
+To support arbitrarily complex parameters, both ResourceClass and
+ResourceClaim contain one field which holds a
+runtime.RawExtension. Validation can be handled by plugins through an
+admission controller (if desired) or later at runtime when the
+parameters are passed to the plugin.
+
+The ResourceClaim spec is read-only once created. The ResourceClaim
+status is reserved for system usage and holds the current state of the
+resource. The status must not get lost. This is departing from how
+Kubernetes traditionally handled status, but something that more
+recently [became more
+acceptable](https://github.com/kubernetes/enhancements/pull/2537). Kube-scheduler
+and plugin communicate by modifying that status. The status is also
+how Kubernetes tracks that a plugin has allocated the resource and on
+which nodes the resource is available.
+
+This approach is an intentional simplification compared to the PV/PVC
+model for volumes because we don't need to deal with two objects when
+allocating resources and therefore don't need something like the
+volume binding controller. If desired, a resource plugin can implement
+support for manually allocated (“static provisioning” in the context
+of volumes) and/or existing resources by reacting to ResourceClaims by
+using those resources to satisfy a claim.
+
+Allocation of a resource happens either immediately (“immediate
+allocation”) or when a Pod needs the resource (“delayed allocation”),
+depending on a flag in the ResourceClaim spec. Pods reference resource
```

Comment:
I understand the design here is to learn PV/PVC.  But can you explain the specific scenario and the reason why we need to allocate resources in advance before the pod is created?

Reply:
Such a resource gets allocated once and then can be used by multiple pods, either in parallel or one after the other.

The user example shows how two containers in the same pod share a GPU allocation. With this API, the same could also be done with two different Pods, if that is what the user wants.

Reusing a resource after one Pod is done with it may be useful when allocating the resource is very expensive, for example programming an FPGA.

Reply:
I added a section about this.

Reply:
How do two pods point to the same resource? Each pod has its own ResourceClaim, right?

Reply:
Both pods refer to the same ResourceClaim by name. That ResourceClaim then needs to be created separately from both pods instead of relying on the automatic creation via inline ResourceClaimTemplates.